File: uganda_compliance/efris/page/efris_diagnostics/efris_diagnostics.py
"""Whitelisted backends for the EFRIS Diagnostics page.

The page renders four panels:

- **Overview** — sync-job counts by status (last 24h) and per-interface success rate.
- **Circuit breakers** — current breaker state per (company, interfaceCode).
- **AES sessions** — for each enabled company, whether an AES key is cached.
- **Recent failures** — last 20 failed sync-jobs / non-success request logs.

Everything is read-only; mutations go through the underlying doctypes.
"""
import time

import frappe
from frappe.utils import add_to_date, now_datetime

from uganda_compliance.efris.client import session as session_cache


@frappe.whitelist()
def get_overview() -> dict:
    """Aggregate sync-job stats over the last 24h."""
    since = add_to_date(now_datetime(), hours=-24)
    rows = frappe.db.sql(
        """
        SELECT status, COUNT(*) AS n
        FROM `tabEFRIS Sync Job`
        WHERE creation >= %s
        GROUP BY status
        """,
        (since,),
        as_dict=True,
    )
    by_status = {r["status"]: r["n"] for r in rows}

    interface_rows = frappe.db.sql(
        """
        SELECT
            interface_code,
            SUM(CASE WHEN status = 'Synced' THEN 1 ELSE 0 END) AS synced,
            SUM(CASE WHEN status = 'Failed' THEN 1 ELSE 0 END) AS failed,
            SUM(CASE WHEN status IN ('Pending', 'In Progress') THEN 1 ELSE 0 END) AS pending,
            COUNT(*) AS total
        FROM `tabEFRIS Sync Job`
        WHERE creation >= %s
        GROUP BY interface_code
        ORDER BY total DESC
        """,
        (since,),
        as_dict=True,
    )
    for row in interface_rows:
        total = row["total"] or 0
        row["success_rate"] = round((row["synced"] / total) * 100, 1) if total else None

    return {
        "window_hours": 24,
        "by_status": by_status,
        "by_interface": interface_rows,
    }


@frappe.whitelist()
def get_breaker_states() -> list:
    """Inspect the circuit breaker state held in `frappe.cache` for every enabled company."""
    companies = [
        r["company"]
        for r in frappe.get_all(
            "E Invoicing Settings",
            filters={"enabled": 1},
            fields=["company"],
        )
    ]
    interfaces = ("T107", "T108", "T109", "T110", "T111", "T119", "T121", "T130", "T131", "T144")
    rows = []
    for company in companies:
        for interface in interfaces:
            raw = frappe.cache().get_value(f"efris:breaker:{company}:{interface}")
            if not raw:
                continue
            try:
                if isinstance(raw, bytes):
                    raw = raw.decode("utf-8")
                state = frappe.parse_json(raw) if isinstance(raw, str) else raw
            except Exception:
                continue
            opened_at = state.get("opened_at") or 0
            opened_age = int(time.time() - opened_at) if opened_at else None
            rows.append(
                {
                    "company": company,
                    "interface_code": interface,
                    "state": state.get("state"),
                    "failures": state.get("failures"),
                    "opened_age_seconds": opened_age,
                }
            )
    return rows


@frappe.whitelist()
def get_session_keys() -> list:
    """For each enabled company, report whether an AES key is currently cached."""
    settings = frappe.get_all(
        "E Invoicing Settings",
        filters={"enabled": 1},
        fields=["company", "tin", "device_no", "sandbox_mode", "sandbox_portal_url", "live_portal_url"],
    )
    rows = []
    for s in settings:
        mode_url = s.get("sandbox_portal_url") if s.get("sandbox_mode") else s.get("live_portal_url")
        cached = session_cache.get_cached(s.get("tin"), s.get("device_no"), mode_url or "")
        rows.append(
            {
                "company": s.get("company"),
                "tin": s.get("tin"),
                "device_no": s.get("device_no"),
                "sandbox_mode": s.get("sandbox_mode"),
                "key_cached": bool(cached),
            }
        )
    return rows


@frappe.whitelist()
def get_recent_failures(limit: int = 20) -> list:
    rows = frappe.get_all(
        "EFRIS Sync Job",
        filters={"status": ["in", ("Failed", "Pending")]},
        fields=[
            "name",
            "company",
            "interface_code",
            "status",
            "attempts",
            "max_attempts",
            "last_attempt",
            "next_attempt_at",
            "reference_doctype",
            "reference_name",
            "left(last_error, 300) as last_error",
        ],
        order_by="modified desc",
        limit=limit,
    )
    return rows


@frappe.whitelist()
def reset_breaker(company: str, interface_code: str) -> dict:
    """Manually clear a tripped breaker (audit log gets the user)."""
    frappe.only_for(("System Manager", "Accounts Manager"))
    frappe.cache().delete_value(f"efris:breaker:{company}:{interface_code}")
    return {"ok": True}


@frappe.whitelist()
def invalidate_session_key(company: str) -> dict:
    """Drop the cached AES key so the next call re-handshakes via T104."""
    frappe.only_for(("System Manager", "Accounts Manager"))
    s = frappe.get_value(
        "E Invoicing Settings",
        {"company": company, "enabled": 1},
        ["tin", "device_no", "sandbox_mode", "sandbox_portal_url", "live_portal_url"],
        as_dict=True,
    )
    if not s:
        return {"ok": False, "error": "No enabled E Invoicing Settings for company"}
    mode_url = s.get("sandbox_portal_url") if s.get("sandbox_mode") else s.get("live_portal_url")
    session_cache.invalidate(s.get("tin"), s.get("device_no"), mode_url or "")
    return {"ok": True}

File: uganda_compliance/efris/page/efris_diagnostics/efris_diagnostics.js
frappe.pages["efris-diagnostics"].on_page_load = function (wrapper) {
    const page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __("EFRIS Diagnostics"),
        single_column: true,
    });

    const $body = $(`
        <div class="efris-diagnostics" style="padding: 12px;">
            <div class="row">
                <div class="col-md-6">
                    <h4>${__("Last 24h")}</h4>
                    <div class="overview-status text-muted">${__("Loading…")}</div>
                    <table class="table table-sm interface-table" style="margin-top: 8px;">
                        <thead><tr>
                            <th>${__("Interface")}</th>
                            <th class="text-right">${__("Total")}</th>
                            <th class="text-right">${__("Synced")}</th>
                            <th class="text-right">${__("Failed")}</th>
                            <th class="text-right">${__("Pending")}</th>
                            <th class="text-right">${__("Success %")}</th>
                        </tr></thead>
                        <tbody></tbody>
                    </table>
                </div>
                <div class="col-md-6">
                    <h4>${__("AES Sessions")}</h4>
                    <table class="table table-sm sessions-table">
                        <thead><tr>
                            <th>${__("Company")}</th>
                            <th>${__("TIN")}</th>
                            <th>${__("Device")}</th>
                            <th>${__("Mode")}</th>
                            <th>${__("Cached")}</th>
                            <th></th>
                        </tr></thead>
                        <tbody></tbody>
                    </table>
                </div>
            </div>

            <h4 style="margin-top: 24px;">${__("Circuit Breakers")}</h4>
            <table class="table table-sm breakers-table">
                <thead><tr>
                    <th>${__("Company")}</th>
                    <th>${__("Interface")}</th>
                    <th>${__("State")}</th>
                    <th class="text-right">${__("Failures")}</th>
                    <th class="text-right">${__("Open age (s)")}</th>
                    <th></th>
                </tr></thead>
                <tbody></tbody>
            </table>

            <h4 style="margin-top: 24px;">${__("Recent Failures")}</h4>
            <table class="table table-sm failures-table">
                <thead><tr>
                    <th>${__("Job")}</th>
                    <th>${__("Company")}</th>
                    <th>${__("Interface")}</th>
                    <th>${__("Status")}</th>
                    <th class="text-right">${__("Attempts")}</th>
                    <th>${__("Reference")}</th>
                    <th>${__("Last Error")}</th>
                </tr></thead>
                <tbody></tbody>
            </table>
        </div>
    `).appendTo(page.body);

    page.set_primary_action(__("Refresh"), () => refresh(), "refresh");

    function refresh() {
        load_overview();
        load_breakers();
        load_sessions();
        load_failures();
    }

    function call(method, args, cb) {
        frappe.call({
            method: `uganda_compliance.efris.page.efris_diagnostics.efris_diagnostics.${method}`,
            args: args || {},
            callback: r => cb(r.message),
        });
    }

    function load_overview() {
        call("get_overview", null, data => {
            const status_text = Object.entries(data.by_status || {})
                .map(([k, v]) => `${k}: ${v}`)
                .join(" · ") || __("No jobs in window");
            $body.find(".overview-status").text(status_text);
            const $tbody = $body.find(".interface-table tbody").empty();
            (data.by_interface || []).forEach(r => {
                $tbody.append(`<tr>
                    <td><code>${r.interface_code}</code></td>
                    <td class="text-right">${r.total}</td>
                    <td class="text-right">${r.synced}</td>
                    <td class="text-right text-danger">${r.failed}</td>
                    <td class="text-right">${r.pending}</td>
                    <td class="text-right">${r.success_rate ?? "—"}</td>
                </tr>`);
            });
        });
    }

    function load_breakers() {
        call("get_breaker_states", null, rows => {
            const $tbody = $body.find(".breakers-table tbody").empty();
            if (!rows || !rows.length) {
                $tbody.append(`<tr><td colspan="6" class="text-muted">${__("All breakers CLOSED")}</td></tr>`);
                return;
            }
            rows.forEach(r => {
                const cls = r.state === "OPEN" ? "text-danger" : (r.state === "HALF_OPEN" ? "text-warning" : "");
                $tbody.append(`<tr>
                    <td>${frappe.utils.escape_html(r.company)}</td>
                    <td><code>${r.interface_code}</code></td>
                    <td class="${cls}">${r.state}</td>
                    <td class="text-right">${r.failures || 0}</td>
                    <td class="text-right">${r.opened_age_seconds ?? "—"}</td>
                    <td><button class="btn btn-xs btn-default reset-breaker"
                        data-company="${frappe.utils.escape_html(r.company)}"
                        data-interface="${r.interface_code}">${__("Reset")}</button></td>
                </tr>`);
            });
        });
    }

    function load_sessions() {
        call("get_session_keys", null, rows => {
            const $tbody = $body.find(".sessions-table tbody").empty();
            (rows || []).forEach(r => {
                $tbody.append(`<tr>
                    <td>${frappe.utils.escape_html(r.company)}</td>
                    <td>${frappe.utils.escape_html(r.tin || "")}</td>
                    <td>${frappe.utils.escape_html(r.device_no || "")}</td>
                    <td>${r.sandbox_mode ? __("Sandbox") : __("Live")}</td>
                    <td>${r.key_cached ? "✓" : "—"}</td>
                    <td><button class="btn btn-xs btn-default invalidate-key"
                        data-company="${frappe.utils.escape_html(r.company)}">${__("Invalidate")}</button></td>
                </tr>`);
            });
        });
    }

    function load_failures() {
        call("get_recent_failures", { limit: 20 }, rows => {
            const $tbody = $body.find(".failures-table tbody").empty();
            if (!rows || !rows.length) {
                $tbody.append(`<tr><td colspan="7" class="text-muted">${__("No recent failures")}</td></tr>`);
                return;
            }
            rows.forEach(r => {
                const ref = r.reference_doctype && r.reference_name
                    ? `<a href="/app/${frappe.router.slug(r.reference_doctype)}/${encodeURIComponent(r.reference_name)}">${frappe.utils.escape_html(r.reference_name)}</a>`
                    : "";
                $tbody.append(`<tr>
                    <td><a href="/app/efris-sync-job/${r.name}">${r.name}</a></td>
                    <td>${frappe.utils.escape_html(r.company)}</td>
                    <td><code>${r.interface_code}</code></td>
                    <td>${r.status}</td>
                    <td class="text-right">${r.attempts}/${r.max_attempts}</td>
                    <td>${ref}</td>
                    <td><small>${frappe.utils.escape_html(r.last_error || "")}</small></td>
                </tr>`);
            });
        });
    }

    $body.on("click", ".reset-breaker", function () {
        const company = $(this).data("company");
        const interface_code = $(this).data("interface");
        call("reset_breaker", { company, interface_code }, () => {
            frappe.show_alert({ message: __("Breaker reset"), indicator: "green" });
            load_breakers();
        });
    });

    $body.on("click", ".invalidate-key", function () {
        const company = $(this).data("company");
        call("invalidate_session_key", { company }, () => {
            frappe.show_alert({ message: __("AES key invalidated"), indicator: "green" });
            load_sessions();
        });
    });

    refresh();
};
